// @telegram-apps/init-data-node/dist/entries/node.js
import { createHmac as createHmac$1 } from "node:crypto";
import { h as hashToken$1, i as isValid$1, s as sign$1, a as signData$1, v as validate$1 } from "./parse-B0zgrTOJ.js";
import { e, f, E, b, g, S, d, c, F, p } from "./parse-B0zgrTOJ.js";
const createHmac = (data, key) => {
  return createHmac$1("sha256", key).update(data).digest();
};
function hashToken(token) {
  return hashToken$1(token, createHmac);
}
function isValid(value, token, options) {
  return isValid$1(value, token, validate, options);
}
function sign(data, key, authDate, options) {
  return sign$1(data, key, authDate, signData, options);
}
function signData(data, key, options) {
  return signData$1(data, key, createHmac, options);
}
function validate(value, token, options) {
  return validate$1(value, token, signData, options);
}
export {
  e as ERR_AUTH_DATE_INVALID,
  f as ERR_EXPIRED,
  E as ERR_HASH_INVALID,
  b as ERR_PARSE,
  g as ERR_SIGN_INVALID,
  S as ERR_UNEXPECTED_VALUE,
  d as TypedError,
  hashToken,
  c as initDataToSearchParams,
  F as isErrorOfType,
  isValid,
  p as parse,
  sign,
  signData,
  validate
};
