// @telegram-apps/init-data-node/dist/entries/web.js
import { h as hashToken$1, i as isValid$1, s as sign$1, a as signData$1, v as validate$1 } from "./parse-B0zgrTOJ.js";
import { e, f, E, b, g, S, d, c, F, p } from "./parse-B0zgrTOJ.js";
const createHmac = async (data, key) => {
  const encoder = new TextEncoder();
  return Buffer.from(
    await crypto.subtle.sign(
      "HMAC",
      await crypto.subtle.importKey(
        "raw",
        typeof key === "string" ? encoder.encode(key) : key,
        { name: "HMAC", hash: "SHA-256" },
        false,
        ["sign", "verify"]
      ),
      encoder.encode(data.toString())
    )
  );
};
function hashToken(token) {
  return hashToken$1(token, createHmac);
}
function isValid(value, token, options) {
  return isValid$1(value, token, validate, options);
}
function sign(data, key, authDate, options) {
  return sign$1(data, key, authDate, signData, options);
}
function signData(data, key, options) {
  return signData$1(data, key, createHmac, options);
}
function validate(value, token, options) {
  return Promise.resolve().then(() => validate$1(value, token, signData, options));
}
export {
  e as ERR_AUTH_DATE_INVALID,
  f as ERR_EXPIRED,
  E as ERR_HASH_INVALID,
  b as ERR_PARSE,
  g as ERR_SIGN_INVALID,
  S as ERR_UNEXPECTED_VALUE,
  d as TypedError,
  hashToken,
  c as initDataToSearchParams,
  F as isErrorOfType,
  isValid,
  p as parse,
  sign,
  signData,
  validate
};
